### DISCUSSION TITLE
Contribution and the future of LUYA update compatibility with angular

### DISCUSSION THREAD

**Discussion by andrewpros** [luyadev/luya#2181] [Questions & Problems]
Hello.

I wanted to ask if the dev env project https://github.com/luyadev/luya-env-dev is still the recommended workflow to update/make pull requests to the luya core repos?

Its connected to the other part of the discussion, outdated client libs, angularjs and ui router used by LUYA, mostly as a base in admin module and every module using it, but mainly also the cms module.

To put it short, are there any plans, ideas, time frame or discussions on how, when or in what form luya will update the outdated client libs going forward?

I had already some discussion in part about it due to my older issue and some tinkering in the internals here https://github.com/luyadev/luya/issues/2122

Now while it will still work for a long time, at some point there would need to be some updates to address it, the current version of angularjs is 1.8.x and very old ui-router 0.4.3, they are just not only old and outdated, they are basically end of life products, dead forever.

Im not even sure there is a direct update path, probably not as the current angular is way different than they way its used in luya. Maybe other libs would be more suitable.

Im asking all of this cuz, its nice to have a discussion about it, im not asking for some detailed plan with dates, just a casual talk about what are the thought on that matter as we all know that one day, there will be some updates needed to the project.

And also i still have that project of mine i want to support, i may consider even some PRs, its probably possible to update at least the ui-router to 1.x which is almost the same from what i have seen, but gives much more control and accessibility with more advanced tinkering, thats my possible goal no1, cuz at least that seems to be the culprit 101 using the admin module the way it is and still having your own things on top of that. Seems like something with minimal admin and cms module updates needed.

Updating or more like migrating the angularjs part would be ofc much harder process with a need to a lot of changes to the admin/cms modules, so out of question right now, but maybe it would be not even needed, not sure, but i will leave here for now.

**Comment by nadar** [comment]
Dear @andrewpros 

Thanks for the question and also the interest in having such a discussion.

>  https://github.com/luyadev/luya-env-dev is still the recommended workflow

We still use that, because it put together all parts of luya, makes it more easy when working with all those modules and core library

> probably not as the current angular is way different than they way its used in luya. Maybe other libs would be more suitable.

Thats indeed one of the major problems. I have explained already once the pro's of angularjs, and why its not possible to migrate to any other library. => https://github.com/luyadev/luya/issues/1824#issuecomment-398297950

> And also i still have that project of mine i want to support, i may consider even some PRs, its probably possible to update at least the ui-router to 1.x

We will update to latest patch releases whenever possible and as long as we can, of course. We have hundreds of sites developed with LUYA and customer and developers are still happy, because it's always receives update trough vendor and not a plugin system. its very stable :-) 

> Seems like something with minimal admin and cms module updates needed.

So if there is someone who d'like to update to latest angular router we can always release a new major version and we are always accepting PR's :-)

**Comment by andrewpros** [comment]
Ah, sorry, i totally forgot about this.

I did finally setup the luya-env-dev and it worked fine, but it had some issues, im sure i did write them somewhere, but cant find it now.

I remember for starters the setup env config file was missing or wrong name, so it was not automatic.
There was some issue that i could not download one repo with the setup configuration, it had to be all 3 of them.
Maybe some other minor issues, but i finally made it work.

Then i found out the current bower angular ui router dependency has no upgrade path, the upgrade version on bower chain is just empty, literally, no files, and there is no latest angular ui router there. Its been a while to maybe its fixed now but not sure.

Anyway, i had to use different package, i replaced "bower-asset/ui-router": "^0.4.0" to  "bower-asset/angular-ui-router": "^1.0.30"

Minor changes were needed, they are here https://github.com/andrewpros/luya-module-admin/tree/dev-updates

It just worked out of the box then.

I did tinker with it more, check other things, ultimately it is compatible with current luya as a drop in replacement if that is the desire, with simple changes above.

BUT

Also ultimately, while it is newer, latest, it does not provide much more than the current possibilities, i can do the same in the current ui router, add my custom routes anyway, turned out, latest version dont make it any simpler, so not sure there is a point, there is also not much new things of any use.

So overall, while possible, its not necessary worth the effort.

Taking into consideration that luya is heavily based on the very convenient way the classic angular works and that there is really no similar upgrade path, it would be very difficult to upgrade anything beyond that latest ui router, especially considering the need for on going support for hundreds of luya projects.

For bigger upgrades and changes probably a new LUYA 2.0 iteration, if ever possible, is the only way to go.
